### DISCUSSION TITLE
read SINT8 and UINT8 using client.convert_from_registers(...) ?

### DISCUSSION THREAD

**Discussion by ramo48859** [pymodbus-dev/pymodbus#2520] [Help] [answered]
How to read an SIN8 correctly using convert_from_registers since BinaryPayloadDecoder is deprecated?

I cannot find any examples...
Just specifying the type as client.DATATYPE.INT8 fails because INT8 is not element of the class DATATYPE?

How to do that?

BR,
David

**Comment by janiversen** [comment, marked as answer]
What is SINT8 ?

The modbus protocol do not support less than 16bit (if you read e.g. 8 coils==bit the response is padded to 16bit bit.

So an INT8 cannot be read, it will be padded as an INT16.

If you have (a really unique and weird) modbus device that implements int8 then you have to see with debug how it's actually transmitted because it's surely not modbus standard.

If you device packs 2 int8 in one register (again something very unusual), then you run a to_bytes on the register returned.

to/from is there to help with the standard cases, more advanced usage (as documented) is left to the app.

**Reply by qcda1** [reply]
Thanks for the reply. Back in the early 80s on old Modicon PLC installations and BCD counter disks with 4 contacts, you could have the four digits in a holding register. The infor was read by a Foxboro AI process control computer using Modbus. Memory was tight at that time.
In the PLCs, you would also deal with bytes when building messages in ASCII for the HMIs. So apart from the fact the holding registers memory map of the PLCs are in 16bits words, they can can be managed at the bit, byte and word levels if I recall.

I have the same challenge with a Midnite Solar Classic 150 solar charge controller that stores operating data in it's memory map based on Modicon register schema and communication protocol. Here's a sample program using BinaryPayloadDecoder. Now, for this manufacturer to choose a standard for its communication is a great plus and should be encouraged compared to the others that tries to reinvent the wheel with weird protocols.

Thanks again and best regards and whishes for the new year. Your code is very well appreciated.

# --------------------------------------------------------------------------- # 
# Based on the address, return the decoded OrderedDict
# --------------------------------------------------------------------------- # 
def doDecode(addr, decoder):
    if (addr == 4100 ):
        decoded = OrderedDict([
            ('PCB', decoder.decode_8bit_uint()),                       #4101 MSB
            ('Type', decoder.decode_8bit_uint()),                      #4101 LSB
            ('Year', decoder.decode_16bit_uint()),                     #4102
            ('Month', decoder.decode_8bit_uint()),                     #4103 MSB
            ('Day', decoder.decode_8bit_uint()),                       #4103 LSB
            ('InfoFlagBits3', decoder.decode_16bit_uint()),            #4104
            ('ignore', decoder.skip_bytes(2)),                         #4105 Reserved
            ('mac_1', decoder.decode_8bit_uint()),                     #4106 MSB  
            ('mac_0', decoder.decode_8bit_uint()),                     #4106 LSB
            ('mac_3', decoder.decode_8bit_uint()),                     #4107 MSB
            ('mac_2', decoder.decode_8bit_uint()),                     #4107 LSB
            ('mac_5', decoder.decode_8bit_uint()),                     #4108 MSB
            ('mac_4', decoder.decode_8bit_uint()),                     #4108 LSB
            ('ignore2', decoder.skip_bytes(4)),                        #4109, 4110
            ('unitID', decoder.decode_32bit_uint()),                   #4111
            ('StatusRoll', decoder.decode_16bit_uint()),               #4113
            ('RsetTmms', decoder.decode_16bit_uint()),                 #4114
            ('BatVoltage', decoder.decode_16bit_int()/10.0),           #4115
            ('PVVoltage', decoder.decode_16bit_uint()/10.0),           #4116
            ('BatCurrent', decoder.decode_16bit_uint()/10.0),          #4117
            ('EnergyToday', decoder.decode_16bit_uint()/10.0),         #4118
            ('Power', decoder.decode_16bit_uint()/1.0),                #4119
            ('ChargeStage', decoder.decode_8bit_uint()),               #4120 MSB
            ('State', decoder.decode_8bit_uint()),                     #4120 LSB
            ('PVCurrent', decoder.decode_16bit_uint()/10.0),           #4121
            ('lastVOC', decoder.decode_16bit_uint()/10.0),             #4122
            ('HighestVinputLog', decoder.decode_16bit_uint()),         #4123
            ('MatchPointShadow', decoder.decode_16bit_uint()),         #4124
            ('AmpHours', decoder.decode_16bit_uint()),                 #4125
            ('TotalEnergy', decoder.decode_32bit_uint()/10.0),         #4126, 4127
            ('LifetimeAmpHours', decoder.decode_32bit_uint()),         #4128, 4129
            ('InfoFlagsBits', decoder.decode_32bit_uint()),            #4130, 31
            ('BatTemperature', decoder.decode_16bit_int()/10.0),       #4132
            ('FETTemperature', decoder.decode_16bit_int()/10.0),       #4133
            ('PCBTemperature', decoder.decode_16bit_int()/10.0),       #4134
            ('NiteMinutesNoPwr', decoder.decode_16bit_uint()),         #4135
            ('MinuteLogIntervalSec', decoder.decode_16bit_uint()),     #4136
            ('modbus_port_register', decoder.decode_16bit_uint()),     #4137
            ('FloatTimeTodaySeconds', decoder.decode_16bit_uint()),    #4138
            ('AbsorbTime', decoder.decode_16bit_uint()),               #4139
            ('reserved1', decoder.decode_16bit_uint()),                #4140
            ('PWM_ReadOnly', decoder.decode_16bit_uint()),             #4141
            ('Reason_For_Reset', decoder.decode_16bit_uint()),         #4142
            ('EqualizeTime', decoder.decode_16bit_uint()),             #4143
        ])

**Reply by janiversen** [reply]
As far as I can see you pack 2 int8 in 1 register, just use to_bytes (standard method) to split the register.

**Reply by qcda1** [reply]
Thanks again for your reply. I'm still a not bad beginner in Python. I mainly used RPG and C on AS/400s in the 80s-90s. Developed a Modicon Modbus protocol driver and Allen&Bradley DataHighway+ specifications for the AS/400 in C/400 so it could control a wrapline in a pulp and manufacturing plant.

"to_byte" will certainly work!

Regards,

Daniel.
